src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -2421,166 +2411,72 @@ public int EndReceiveFrom(IAsyncResult asyncResult, ref EndPoint endPoint)
             return result.ReceivedBytes;
         }
 
-        // Routine Description:
-        //
-        //    BeginAccept - Does an async winsock accept, creating a new socket on success
-        //
-        //     Works by creating a pending accept request the first time,
-        //     and subsequent calls are queued so that when the first accept completes,
-        //     the next accept can be resubmitted in the callback.
-        //     this routine may go pending at which time,
-        //     but any case the callback Delegate will be called upon completion
-        //
-        // Arguments:
-        //
-        //    Callback - Async Callback Delegate that is called upon Async Completion
-        //    State - State used to track callback, set by caller, not required
-        //
-        // Return Value:
-        //
-        //    IAsyncResult - Async result used to retrieve resultant new socket
-        public IAsyncResult BeginAccept(AsyncCallback? callback, object? state)
-        {
-            if (!_isDisconnected)
-            {
-                return BeginAcceptCommon(acceptSocket: null, receiveSize: 0, callback, state);
-            }
-
-            Debug.Assert(Disposed);
-            ThrowObjectDisposedException();
-            return null; // unreachable
-        }
+        public IAsyncResult BeginAccept(AsyncCallback? callback, object? state) =>
+            TaskToApmBeginWithSyncExceptions(AcceptAsync(), callback, state);
```

Comment:
When was this added? I don't remember seeing it.  Is this actually providing compatibility with what was there before? It's not just throwing exceptions that may have occurred on the initiation path,  it's throwing all exceptions from the whole operation if it happens to complete quickly, by the time we get to the IsFaulted check.  I don't have a strong opinion,  but if our goal is compatibility, i think this likely makes it worse, potentially throwing new exceptions that previously weren't possible from Begin. I suggest we just accept the small difference in behavior we've accepted everywhere else we've used TaskToApm, rather than trying to special-case this.

Reply:
This got added in https://github.com/dotnet/runtime/pull/51213 (which I just merged last night) to address the compat concerns from @antonfirsov. I didn't think it would be controversial, so I merged it without additional review.

It provides compat with what was there before in the sense that sync exceptions now propagate synchronously. It also, as you point out, will throw synchronously if the task completes asynchronously but before we hit the IsFaulted check. AFAIK there is no way to avoid this (right?).

> I suggest we just accept the small difference in behavior we've accepted everywhere else we've used TaskToApm, rather than trying to special-case this. 

I'm fine with this. I'll just rip this code out. @antonfirsov any concerns here? Based on this I think we should close #47905 without action as well.

Reply:
@stephentoub le me know, if my definition of breaking changes, and/or my expectations around maintaining docs are too strict, or if I'm misunderstanding the rules in other means, but this is how I view it:

AFAIK we do not document async exceptions the same way we do with we sync exceptions since they are not being actually thrown by calling the method, but rather by awaiting it.

This means that if we close #47905 as wontfix, we need to add a "breaking change" label to all these PR-s, and make sure we propagate and address all the related documentation work in 6.0. If the goal is to save time, I'm don't think this path will result in less work. I would just add a tiny layer of compat code instead.

Reply:
Whether the behavior was documented or not, it can still be a breaking change. And my point is that in this case, it's a breaking change whether or not this TaskToApmBeginWithSyncExceptions is used, and from my perspective, it's a larger break with TaskToApmBeginWithSyncExceptions (it's non-deterministically causing more exceptions to be thrown out of the BeginXx method).  I'm not convinced either way it rises to the level that requires being documented as a breaking change, but if you feel it needs to be, feel free.  To my knowledge we haven't in the past when we've ripped out APM implementations and replaced them with TaskToApm wrappers around XxAsync methods.

Reply:
Before I answer, just to make sure we're on the same page, you agree that using TaskToApmBeginWithSyncExceptions also changed the exceptions that could come from Begin, yes?

Reply:
`TaskToApmBeginWithSyncExceptions` yes, I'm not arguing pro that particular change, I think there are other ways to keep compatibility. I just want to see us making a conscious decision around this topic.

Reply:
@antonfirsov How would you maintain compat here?

There is a fundamental issue here, which is that the Task APIs do not expose synchronous completion in the same way that the old APM APIs do. As such we are either going to report some exceptions from Begin that used to come from End (as with my TaskToApmBeginWithSyncExceptions), or vice versa.

Reply:
OK, I falsely assumed there is a simple way (or cheap workaround) to detect synchronous completions in `Task` and `ValueTask`. I realized it's not possible, so maintaining compatibility is more complicated here than I originally thought.

We can still accomplish it with a trick like the one below, although I understand if it's not the way we want to go. In that case I recommend to open an issue in `dotnet-api-docs` for #43845 instead of fixing #47905.

```C#
internal class FailedSynchronouslyCompletionSource : TaskCompletionSource
{
    public FailedSynchronouslyCompletionSource(Exception exception) : base(exception)
    {
        SetException(exception);
    }
}

private static Task TaskFromSynchrnonousException(Exception ex)
{
    return new FailedSynchronouslyCompletionSource(ex).Task;
}

private static IAsyncResult TaskToApmBeginWithSyncExceptions(Task task, AsyncCallback? callback, object? state)
{
    if (task.IsFaulted && task.AsyncState is Exception ex)
    {
        throw ex;
    }

    return TaskToApm.Begin(task, callback, state);
}
```

Reply:
> We can still accomplish it with a trick like the one below

Can you share a full example?  I think it's much more complicated than that, and would deoptimize the XxAsync methods we're trying to reuse.

Reply:
Opened #51693 to show a more complete example.

Reply:
Thanks.  I left a few comments.  That only partially addresses the issue, and adds non-trivial complication.  I don't believe it's worth it.  If you believe this rises to the level of requiring a breaking change notification, please feel free to do so.  But FYI we've made such changes throughout .NET Core incrementally over the last several years, ripping out lots of custom IAsyncResult code across lots of libraries, and I don't believe we've documented any of them as being breaking.

Reply:
Thanks! I see now the more complicated cases, and agree with your analysis, that it's not worth it.

> ripping out lots of custom IAsyncResult code across lots of libraries, and I don't believe we've documented any of them as being breaking.

When it comes to the process that is triggered by applying the "breaking change" labels today, I don't think it makes sense to deviate from an established practice in this particular case of sockets. I opened #6658 to track the API docs work.

Reply:
> I opened #6658 to track the API docs work.

This looks like it's linked to the wrong repo, perhaps?